### PR TITLE
Externalize more dataflow UI strings

### PR DESCRIPTION
**Pull request by elharo**
@tgroh

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/MessagesTest.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2016 Google Inc.
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/MessagesTest.java]: 2017

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/page/MessagesTest.java [hunk 2]
```diff
@@ -19,6 +19,8 @@
 import org.junit.Assert;
 import org.junit.Test;
 
+import com.google.cloud.tools.eclipse.dataflow.ui.Messages;
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/page/MessagesTest.java]: Why don't you move the tests in this class to `.dataflow.ui.Messages` you just added here, and then delete this class?

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/MessagesTest.java (hunk 1 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/MessagesTest.java]: done

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/page/MessagesTest.java (hunk 2 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/page/MessagesTest.java]: done

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/MessagesTest.java [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016 Google Inc.
+ * Copyright 2017 Google Inc.
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/MessagesTest.java]: Don't need to update the copyright year.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/MessagesTest.java [hunk 4]
```diff
@@ -14,14 +14,19 @@
...
+package com.google.cloud.tools.eclipse.dataflow.ui;
 
 import org.junit.Assert;
 import org.junit.Test;
 
 public class MessagesTest {
 
   @Test
+  public void testNewCloudDataflowProject() {
+    Assert.assertEquals("New Cloud Dataflow Project", Messages.getString("new.cloud.dataflow.project"));
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/MessagesTest.java]: I think this would fail if it was run on a non-english locale where there were translations.  A hypothetical problem at this point!

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java [hunk 5]
```diff
@@ -311,12 +314,13 @@ public void initializeFrom(ILaunchConfiguration configuration) {
       defaultOptionsComponent.setCustomValues(launchConfiguration.getArgumentValues());
 
       String userOptionsName = launchConfiguration.getUserOptionsName();
-      userOptionsSelector.setText(userOptionsName == null ? "" : userOptionsName);
+      userOptionsSelector.setText(userOptionsName == null ? "" : userOptionsName); //$NON-NLS-1$
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java]: Could use `Strings.nullToEmpty()`

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java [hunk 6]
```diff
@@ -342,7 +346,7 @@ void updateRunnerButtons(MajorVersion majorVersion) {
    * Asynchronously updates the project hierarchy.
    */
   private void updateHierarchy(final MajorVersion majorVersion) {
-    Job job = new Job("Update Hierarchy") {
+    Job job = new Job("Update Hierarchy") { //$NON-NLS-1$
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java]: This string likely should be externalized as it is reported on the Progress View.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java [hunk 7]
```diff
@@ -442,13 +446,13 @@ private boolean validateRequiredGroups(MissingRequiredProperties validationFailu
...
-      errorBuilder.append(". Properties satisfying group requirement are ");
+      errorBuilder.append(". Properties satisfying group requirement are "); //$NON-NLS-1$
       Set<String> groupMembers = new HashSet<>();
       for (PipelineOptionsProperty missingProperty : missingGroupEntry.getValue()) {
         groupMembers.add(missingProperty.getName());
       }
       errorBuilder.append(MISSING_GROUP_MEMBER_JOINER.join(groupMembers));
-      errorBuilder.append(".");
+      errorBuilder.append("."); //$NON-NLS-1$
       setErrorMessage(errorBuilder.toString());
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java]: This looks like a user-oriented string that should be externalized and parameterized.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java [hunk 8]
```diff
@@ -459,7 +463,7 @@ private boolean validateRequiredProperties(MissingRequiredProperties validationF
     PipelineOptionsProperty missingProperty =
         Iterables.getFirst(validationFailures.getMissingProperties(), null);
     if (missingProperty != null) {
-      setErrorMessage("Missing required property " + missingProperty.getName());
+      setErrorMessage("Missing required property " + missingProperty.getName()); //$NON-NLS-1$
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java]: This looks like a user-oriented message that should be externalized and parameterized.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsSelectionDialog.java [hunk 9]
```diff
@@ -140,7 +140,7 @@ public String getText(Object element) {
         return null;
       }
       String simpleName = Signature.getSimpleName(element.toString());
-      return String.format("%s - %s", simpleName, element.toString());
+      return String.format("%s - %s", simpleName, element.toString()); //$NON-NLS-1$
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsSelectionDialog.java]: This should still be externalized for R2L languages.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java (hunk 7 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java]: It's trickier though because of the use of the StringBuilder and the for loop.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java (hunk 8 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java]: done

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsSelectionDialog.java (hunk 9 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsSelectionDialog.java]: what is this used for? i.e. should the name of the property be?

Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsSelectionDialog.java]: It's used in an _Open Type_-like dialog to search for PipelineOptions classes. `pipelineoptions.search.typenames`?

screen shot 2017-08-09 at 4 14 16 pm

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java (hunk 7 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java]: I think we can make this work with ICU's `com.ibm.icu.text.ListFormatter`.  ICU is bundled with Eclipse.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsSelectionDialog.java (hunk 9 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsSelectionDialog.java]: Thanks.  Are you sure that should be localized in RTL languages? Will an RTL UI truncate and scroll to the left instead of the right? I'm not really familiar with RTL UIs.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java (hunk 7 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java]: maybe, but that looks like it might require a different messages class. We can do that in a separate PR if it seems important.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsSelectionDialog.java (hunk 9 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsSelectionDialog.java]: You can enable RTL with _Preferences > General > Globalization_ (easier to do from a debug instance).  In this case, I'm honestly not sure as the result will be filling in type names which won't be RTL'd.  I just checked the JDT _Open Type_ dialog and it doesn't do it.  So I'll retract this.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java (hunk 7 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java]: There's no need for ICU's MessageFormat.
```
// "Missing value for group {0}. Properties satisfying group requirement are {1}."
MessageFormat.format("missing.property.value.for.group",
   missingGroupEntry.getKey(),
   ListFormatter.getInstance().format(groupMembers));
```
Produces:
screen shot 2017-08-10 at 2 45 32 pm

That said, I can't actually excercise this codepath with the DirectRunner and DataflowRunners as they don't seem to specify any required grouped properties.  I had to hack up the code instead.
